// _element-plus@1.2.0-beta.4@element-plus/lib/components/progress/src/index.vue_vue_type_template_id_556e268a_lang.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');

const _hoisted_1 = ["aria-valuenow"];
const _hoisted_2 = {
  key: 0,
  class: "el-progress-bar"
};
const _hoisted_3 = {
  key: 0,
  class: "el-progress-bar__innerText"
};
const _hoisted_4 = { viewBox: "0 0 100 100" };
const _hoisted_5 = ["d", "stroke-width"];
const _hoisted_6 = ["d", "stroke", "stroke-linecap", "stroke-width"];
const _hoisted_7 = { key: 0 };
function render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_el_icon = vue.resolveComponent("el-icon");
  return vue.openBlock(), vue.createElementBlock("div", {
    class: vue.normalizeClass(["el-progress", [
      `el-progress--${_ctx.type}`,
      _ctx.status ? `is-${_ctx.status}` : "",
      {
        "el-progress--without-text": !_ctx.showText,
        "el-progress--text-inside": _ctx.textInside
      }
    ]]),
    role: "progressbar",
    "aria-valuenow": _ctx.percentage,
    "aria-valuemin": "0",
    "aria-valuemax": "100"
  }, [
    _ctx.type === "line" ? (vue.openBlock(), vue.createElementBlock("div", _hoisted_2, [
      vue.createElementVNode("div", {
        class: "el-progress-bar__outer",
        style: vue.normalizeStyle({ height: `${_ctx.strokeWidth}px` })
      }, [
        vue.createElementVNode("div", {
          class: vue.normalizeClass([
            "el-progress-bar__inner",
            { "el-progress-bar__inner--indeterminate": _ctx.indeterminate }
          ]),
          style: vue.normalizeStyle(_ctx.barStyle)
        }, [
          (_ctx.showText || _ctx.$slots.default) && _ctx.textInside ? (vue.openBlock(), vue.createElementBlock("div", _hoisted_3, [
            vue.renderSlot(_ctx.$slots, "default", vue.normalizeProps(vue.guardReactiveProps(_ctx.slotData)), () => [
              vue.createElementVNode("span", null, vue.toDisplayString(_ctx.content), 1)
            ])
          ])) : vue.createCommentVNode("v-if", true)
        ], 6)
      ], 4)
    ])) : (vue.openBlock(), vue.createElementBlock("div", {
      key: 1,
      class: "el-progress-circle",
      style: vue.normalizeStyle({ height: `${_ctx.width}px`, width: `${_ctx.width}px` })
    }, [
      (vue.openBlock(), vue.createElementBlock("svg", _hoisted_4, [
        vue.createElementVNode("path", {
          class: "el-progress-circle__track",
          d: _ctx.trackPath,
          stroke: "#e5e9f2",
          "stroke-width": _ctx.relativeStrokeWidth,
          fill: "none",
          style: vue.normalizeStyle(_ctx.trailPathStyle)
        }, null, 12, _hoisted_5),
        vue.createElementVNode("path", {
          class: "el-progress-circle__path",
          d: _ctx.trackPath,
          stroke: _ctx.stroke,
          fill: "none",
          "stroke-linecap": _ctx.strokeLinecap,
          "stroke-width": _ctx.percentage ? _ctx.relativeStrokeWidth : 0,
          style: vue.normalizeStyle(_ctx.circlePathStyle)
        }, null, 12, _hoisted_6)
      ]))
    ], 4)),
    (_ctx.showText || _ctx.$slots.default) && !_ctx.textInside ? (vue.openBlock(), vue.createElementBlock("div", {
      key: 2,
      class: "el-progress__text",
      style: vue.normalizeStyle({ fontSize: `${_ctx.progressTextSize}px` })
    }, [
      vue.renderSlot(_ctx.$slots, "default", vue.normalizeProps(vue.guardReactiveProps(_ctx.slotData)), () => [
        !_ctx.status ? (vue.openBlock(), vue.createElementBlock("span", _hoisted_7, vue.toDisplayString(_ctx.content), 1)) : (vue.openBlock(), vue.createBlock(_component_el_icon, { key: 1 }, {
          default: vue.withCtx(() => [
            (vue.openBlock(), vue.createBlock(vue.resolveDynamicComponent(_ctx.statusIcon)))
          ]),
          _: 1
        }))
      ])
    ], 4)) : vue.createCommentVNode("v-if", true)
  ], 10, _hoisted_1);
}

exports.render = render;
