// _element-plus@1.2.0-beta.4@element-plus/lib/components/date-picker/src/date-picker.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
var dayjs = require('dayjs');
var customParseFormat = require('dayjs/plugin/customParseFormat');
var advancedFormat = require('dayjs/plugin/advancedFormat');
var localeData = require('dayjs/plugin/localeData');
var weekOfYear = require('dayjs/plugin/weekOfYear');
var weekYear = require('dayjs/plugin/weekYear');
var dayOfYear = require('dayjs/plugin/dayOfYear');
var isSameOrAfter = require('dayjs/plugin/isSameOrAfter');
var isSameOrBefore = require('dayjs/plugin/isSameOrBefore');
require('../../time-picker/index.js');
require('./date-picker-com/panel-date-pick.js');
require('./date-picker-com/panel-date-range.js');
require('./date-picker-com/panel-month-range.js');
var datePicker_type = require('./date-picker.type.js');
var panelDateRange_vue_vue_type_script_lang = require('./date-picker-com/panel-date-range.vue_vue_type_script_lang.js');
var panelMonthRange_vue_vue_type_script_lang = require('./date-picker-com/panel-month-range.vue_vue_type_script_lang.js');
var panelDatePick_vue_vue_type_script_lang = require('./date-picker-com/panel-date-pick.vue_vue_type_script_lang.js');
var props = require('../../time-picker/src/common/props.js');
var constant = require('../../time-picker/src/common/constant.js');
var picker_vue_vue_type_script_lang = require('../../time-picker/src/common/picker.vue_vue_type_script_lang.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var dayjs__default = /*#__PURE__*/_interopDefaultLegacy(dayjs);
var customParseFormat__default = /*#__PURE__*/_interopDefaultLegacy(customParseFormat);
var advancedFormat__default = /*#__PURE__*/_interopDefaultLegacy(advancedFormat);
var localeData__default = /*#__PURE__*/_interopDefaultLegacy(localeData);
var weekOfYear__default = /*#__PURE__*/_interopDefaultLegacy(weekOfYear);
var weekYear__default = /*#__PURE__*/_interopDefaultLegacy(weekYear);
var dayOfYear__default = /*#__PURE__*/_interopDefaultLegacy(dayOfYear);
var isSameOrAfter__default = /*#__PURE__*/_interopDefaultLegacy(isSameOrAfter);
var isSameOrBefore__default = /*#__PURE__*/_interopDefaultLegacy(isSameOrBefore);

dayjs__default["default"].extend(localeData__default["default"]);
dayjs__default["default"].extend(advancedFormat__default["default"]);
dayjs__default["default"].extend(customParseFormat__default["default"]);
dayjs__default["default"].extend(weekOfYear__default["default"]);
dayjs__default["default"].extend(weekYear__default["default"]);
dayjs__default["default"].extend(dayOfYear__default["default"]);
dayjs__default["default"].extend(isSameOrAfter__default["default"]);
dayjs__default["default"].extend(isSameOrBefore__default["default"]);
const getPanel = function(type) {
  if (type === "daterange" || type === "datetimerange") {
    return panelDateRange_vue_vue_type_script_lang["default"];
  } else if (type === "monthrange") {
    return panelMonthRange_vue_vue_type_script_lang["default"];
  }
  return panelDatePick_vue_vue_type_script_lang["default"];
};
var DatePicker = vue.defineComponent({
  name: "ElDatePicker",
  install: null,
  props: {
    ...props.timePickerDefaultProps,
    type: {
      type: String,
      default: "date"
    }
  },
  emits: ["update:modelValue"],
  setup(props, ctx) {
    vue.provide("ElPopperOptions", props.popperOptions);
    vue.provide(datePicker_type.ROOT_PICKER_INJECTION_KEY, {
      ctx
    });
    const commonPicker = vue.ref(null);
    const refProps = {
      ...props,
      focus: () => {
        var _a;
        (_a = commonPicker.value) == null ? void 0 : _a.handleFocus();
      }
    };
    ctx.expose(refProps);
    return () => {
      var _a;
      const format = (_a = props.format) != null ? _a : constant.DEFAULT_FORMATS_DATEPICKER[props.type] || constant.DEFAULT_FORMATS_DATE;
      return vue.h(picker_vue_vue_type_script_lang["default"], {
        ...props,
        format,
        type: props.type,
        ref: commonPicker,
        "onUpdate:modelValue": (value) => ctx.emit("update:modelValue", value)
      }, {
        default: (scopedProps) => vue.h(getPanel(props.type), scopedProps),
        "range-separator": () => vue.renderSlot(ctx.slots, "range-separator")
      });
    };
  }
});

exports["default"] = DatePicker;
