// element-plus/es/components/image-viewer/src/image-viewer.vue_vue_type_template_id_4b22ad85_lang.mjs
import { createElementVNode, resolveComponent, openBlock, createBlock, Transition, withCtx, normalizeStyle, withModifiers, createCommentVNode, createVNode, createElementBlock, Fragment, normalizeClass, resolveDynamicComponent, renderList, withDirectives, vShow, renderSlot } from 'vue';

const _hoisted_1 = { class: "el-image-viewer__btn el-image-viewer__actions" };
const _hoisted_2 = { class: "el-image-viewer__actions__inner" };
const _hoisted_3 = /* @__PURE__ */ createElementVNode("i", { class: "el-image-viewer__actions__divider" }, null, -1);
const _hoisted_4 = /* @__PURE__ */ createElementVNode("i", { class: "el-image-viewer__actions__divider" }, null, -1);
const _hoisted_5 = { class: "el-image-viewer__canvas" };
const _hoisted_6 = ["src"];
function render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_close = resolveComponent("close");
  const _component_el_icon = resolveComponent("el-icon");
  const _component_arrow_left = resolveComponent("arrow-left");
  const _component_arrow_right = resolveComponent("arrow-right");
  const _component_zoom_out = resolveComponent("zoom-out");
  const _component_zoom_in = resolveComponent("zoom-in");
  const _component_refresh_left = resolveComponent("refresh-left");
  const _component_refresh_right = resolveComponent("refresh-right");
  return openBlock(), createBlock(Transition, { name: "viewer-fade" }, {
    default: withCtx(() => [
      createElementVNode("div", {
        ref: "wrapper",
        tabindex: -1,
        class: "el-image-viewer__wrapper",
        style: normalizeStyle({ zIndex: _ctx.zIndex })
      }, [
        createElementVNode("div", {
          class: "el-image-viewer__mask",
          onClick: _cache[0] || (_cache[0] = withModifiers(($event) => _ctx.hideOnClickModal && _ctx.hide(), ["self"]))
        }),
        createCommentVNode(" CLOSE "),
        createElementVNode("span", {
          class: "el-image-viewer__btn el-image-viewer__close",
          onClick: _cache[1] || (_cache[1] = (...args) => _ctx.hide && _ctx.hide(...args))
        }, [
          createVNode(_component_el_icon, null, {
            default: withCtx(() => [
              createVNode(_component_close)
            ]),
            _: 1
          })
        ]),
        createCommentVNode(" ARROW "),
        !_ctx.isSingle ? (openBlock(), createElementBlock(Fragment, { key: 0 }, [
          createElementVNode("span", {
            class: normalizeClass(["el-image-viewer__btn el-image-viewer__prev", { "is-disabled": !_ctx.infinite && _ctx.isFirst }]),
            onClick: _cache[2] || (_cache[2] = (...args) => _ctx.prev && _ctx.prev(...args))
          }, [
            createVNode(_component_el_icon, null, {
              default: withCtx(() => [
                createVNode(_component_arrow_left)
              ]),
              _: 1
            })
          ], 2),
          createElementVNode("span", {
            class: normalizeClass(["el-image-viewer__btn el-image-viewer__next", { "is-disabled": !_ctx.infinite && _ctx.isLast }]),
            onClick: _cache[3] || (_cache[3] = (...args) => _ctx.next && _ctx.next(...args))
          }, [
            createVNode(_component_el_icon, null, {
              default: withCtx(() => [
                createVNode(_component_arrow_right)
              ]),
              _: 1
            })
          ], 2)
        ], 64)) : createCommentVNode("v-if", true),
        createCommentVNode(" ACTIONS "),
        createElementVNode("div", _hoisted_1, [
          createElementVNode("div", _hoisted_2, [
            createVNode(_component_el_icon, {
              onClick: _cache[4] || (_cache[4] = ($event) => _ctx.handleActions("zoomOut"))
            }, {
              default: withCtx(() => [
                createVNode(_component_zoom_out)
              ]),
              _: 1
            }),
            createVNode(_component_el_icon, {
              onClick: _cache[5] || (_cache[5] = ($event) => _ctx.handleActions("zoomIn"))
            }, {
              default: withCtx(() => [
                createVNode(_component_zoom_in)
              ]),
              _: 1
            }),
            _hoisted_3,
            createVNode(_component_el_icon, { onClick: _ctx.toggleMode }, {
              default: withCtx(() => [
                (openBlock(), createBlock(resolveDynamicComponent(_ctx.mode.icon)))
              ]),
              _: 1
            }, 8, ["onClick"]),
            _hoisted_4,
            createVNode(_component_el_icon, {
              onClick: _cache[6] || (_cache[6] = ($event) => _ctx.handleActions("anticlockwise"))
            }, {
              default: withCtx(() => [
                createVNode(_component_refresh_left)
              ]),
              _: 1
            }),
            createVNode(_component_el_icon, {
              onClick: _cache[7] || (_cache[7] = ($event) => _ctx.handleActions("clockwise"))
            }, {
              default: withCtx(() => [
                createVNode(_component_refresh_right)
              ]),
              _: 1
            })
          ])
        ]),
        createCommentVNode(" CANVAS "),
        createElementVNode("div", _hoisted_5, [
          (openBlock(true), createElementBlock(Fragment, null, renderList(_ctx.urlList, (url, i) => {
            return withDirectives((openBlock(), createElementBlock("img", {
              ref: "img",
              key: url,
              src: url,
              style: normalizeStyle(_ctx.imgStyle),
              class: "el-image-viewer__img",
              onLoad: _cache[8] || (_cache[8] = (...args) => _ctx.handleImgLoad && _ctx.handleImgLoad(...args)),
              onError: _cache[9] || (_cache[9] = (...args) => _ctx.handleImgError && _ctx.handleImgError(...args)),
              onMousedown: _cache[10] || (_cache[10] = (...args) => _ctx.handleMouseDown && _ctx.handleMouseDown(...args))
            }, null, 44, _hoisted_6)), [
              [vShow, i === _ctx.index]
            ]);
          }), 128))
        ]),
        renderSlot(_ctx.$slots, "default")
      ], 4)
    ]),
    _: 3
  });
}

export { render };
