// element-plus/lib/components/image-viewer/src/image-viewer.vue_vue_type_template_id_4b22ad85_lang.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');

const _hoisted_1 = { class: "el-image-viewer__btn el-image-viewer__actions" };
const _hoisted_2 = { class: "el-image-viewer__actions__inner" };
const _hoisted_3 = /* @__PURE__ */ vue.createElementVNode("i", { class: "el-image-viewer__actions__divider" }, null, -1);
const _hoisted_4 = /* @__PURE__ */ vue.createElementVNode("i", { class: "el-image-viewer__actions__divider" }, null, -1);
const _hoisted_5 = { class: "el-image-viewer__canvas" };
const _hoisted_6 = ["src"];
function render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_close = vue.resolveComponent("close");
  const _component_el_icon = vue.resolveComponent("el-icon");
  const _component_arrow_left = vue.resolveComponent("arrow-left");
  const _component_arrow_right = vue.resolveComponent("arrow-right");
  const _component_zoom_out = vue.resolveComponent("zoom-out");
  const _component_zoom_in = vue.resolveComponent("zoom-in");
  const _component_refresh_left = vue.resolveComponent("refresh-left");
  const _component_refresh_right = vue.resolveComponent("refresh-right");
  return vue.openBlock(), vue.createBlock(vue.Transition, { name: "viewer-fade" }, {
    default: vue.withCtx(() => [
      vue.createElementVNode("div", {
        ref: "wrapper",
        tabindex: -1,
        class: "el-image-viewer__wrapper",
        style: vue.normalizeStyle({ zIndex: _ctx.zIndex })
      }, [
        vue.createElementVNode("div", {
          class: "el-image-viewer__mask",
          onClick: _cache[0] || (_cache[0] = vue.withModifiers(($event) => _ctx.hideOnClickModal && _ctx.hide(), ["self"]))
        }),
        vue.createCommentVNode(" CLOSE "),
        vue.createElementVNode("span", {
          class: "el-image-viewer__btn el-image-viewer__close",
          onClick: _cache[1] || (_cache[1] = (...args) => _ctx.hide && _ctx.hide(...args))
        }, [
          vue.createVNode(_component_el_icon, null, {
            default: vue.withCtx(() => [
              vue.createVNode(_component_close)
            ]),
            _: 1
          })
        ]),
        vue.createCommentVNode(" ARROW "),
        !_ctx.isSingle ? (vue.openBlock(), vue.createElementBlock(vue.Fragment, { key: 0 }, [
          vue.createElementVNode("span", {
            class: vue.normalizeClass(["el-image-viewer__btn el-image-viewer__prev", { "is-disabled": !_ctx.infinite && _ctx.isFirst }]),
            onClick: _cache[2] || (_cache[2] = (...args) => _ctx.prev && _ctx.prev(...args))
          }, [
            vue.createVNode(_component_el_icon, null, {
              default: vue.withCtx(() => [
                vue.createVNode(_component_arrow_left)
              ]),
              _: 1
            })
          ], 2),
          vue.createElementVNode("span", {
            class: vue.normalizeClass(["el-image-viewer__btn el-image-viewer__next", { "is-disabled": !_ctx.infinite && _ctx.isLast }]),
            onClick: _cache[3] || (_cache[3] = (...args) => _ctx.next && _ctx.next(...args))
          }, [
            vue.createVNode(_component_el_icon, null, {
              default: vue.withCtx(() => [
                vue.createVNode(_component_arrow_right)
              ]),
              _: 1
            })
          ], 2)
        ], 64)) : vue.createCommentVNode("v-if", true),
        vue.createCommentVNode(" ACTIONS "),
        vue.createElementVNode("div", _hoisted_1, [
          vue.createElementVNode("div", _hoisted_2, [
            vue.createVNode(_component_el_icon, {
              onClick: _cache[4] || (_cache[4] = ($event) => _ctx.handleActions("zoomOut"))
            }, {
              default: vue.withCtx(() => [
                vue.createVNode(_component_zoom_out)
              ]),
              _: 1
            }),
            vue.createVNode(_component_el_icon, {
              onClick: _cache[5] || (_cache[5] = ($event) => _ctx.handleActions("zoomIn"))
            }, {
              default: vue.withCtx(() => [
                vue.createVNode(_component_zoom_in)
              ]),
              _: 1
            }),
            _hoisted_3,
            vue.createVNode(_component_el_icon, { onClick: _ctx.toggleMode }, {
              default: vue.withCtx(() => [
                (vue.openBlock(), vue.createBlock(vue.resolveDynamicComponent(_ctx.mode.icon)))
              ]),
              _: 1
            }, 8, ["onClick"]),
            _hoisted_4,
            vue.createVNode(_component_el_icon, {
              onClick: _cache[6] || (_cache[6] = ($event) => _ctx.handleActions("anticlockwise"))
            }, {
              default: vue.withCtx(() => [
                vue.createVNode(_component_refresh_left)
              ]),
              _: 1
            }),
            vue.createVNode(_component_el_icon, {
              onClick: _cache[7] || (_cache[7] = ($event) => _ctx.handleActions("clockwise"))
            }, {
              default: vue.withCtx(() => [
                vue.createVNode(_component_refresh_right)
              ]),
              _: 1
            })
          ])
        ]),
        vue.createCommentVNode(" CANVAS "),
        vue.createElementVNode("div", _hoisted_5, [
          (vue.openBlock(true), vue.createElementBlock(vue.Fragment, null, vue.renderList(_ctx.urlList, (url, i) => {
            return vue.withDirectives((vue.openBlock(), vue.createElementBlock("img", {
              ref: "img",
              key: url,
              src: url,
              style: vue.normalizeStyle(_ctx.imgStyle),
              class: "el-image-viewer__img",
              onLoad: _cache[8] || (_cache[8] = (...args) => _ctx.handleImgLoad && _ctx.handleImgLoad(...args)),
              onError: _cache[9] || (_cache[9] = (...args) => _ctx.handleImgError && _ctx.handleImgError(...args)),
              onMousedown: _cache[10] || (_cache[10] = (...args) => _ctx.handleMouseDown && _ctx.handleMouseDown(...args))
            }, null, 44, _hoisted_6)), [
              [vue.vShow, i === _ctx.index]
            ]);
          }), 128))
        ]),
        vue.renderSlot(_ctx.$slots, "default")
      ], 4)
    ]),
    _: 3
  });
}

exports.render = render;
