// element-plus/lib/components/pagination/src/components/jumper.vue_vue_type_script_lang.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
require('../../../../hooks/index.js');
var index = require('../../../input/index.js');
var usePagination = require('../usePagination.js');
var index$1 = require('../../../../hooks/use-locale/index.js');

var script = vue.defineComponent({
  name: "ElPaginationJumper",
  components: {
    ElInput: index.ElInput
  },
  setup() {
    const { t } = index$1.useLocaleInject();
    const { pageCount, disabled, currentPage, changeEvent } = usePagination.usePagination();
    const userInput = vue.ref();
    const innerValue = vue.computed(() => {
      var _a;
      return (_a = userInput.value) != null ? _a : currentPage == null ? void 0 : currentPage.value;
    });
    function handleInput(val) {
      userInput.value = +val;
    }
    function handleChange(val) {
      changeEvent == null ? void 0 : changeEvent(+val);
      userInput.value = void 0;
    }
    return {
      pageCount,
      disabled,
      innerValue,
      t,
      handleInput,
      handleChange
    };
  }
});

exports["default"] = script;
