// element-plus/lib/components/upload/src/upload-list.vue_vue_type_script_lang.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
var shared = require('@vue/shared');
var index$1 = require('../../icon/index.js');
var icons = require('@element-plus/icons');
require('../../../hooks/index.js');
var index = require('../../progress/index.js');
var index$2 = require('../../../hooks/use-locale/index.js');

var script = vue.defineComponent({
  name: "ElUploadList",
  components: {
    ElProgress: index["default"],
    ElIcon: index$1.ElIcon,
    Document: icons.Document,
    Delete: icons.Delete,
    Close: icons.Close,
    ZoomIn: icons.ZoomIn,
    Check: icons.Check,
    CircleCheck: icons.CircleCheck
  },
  props: {
    files: {
      type: Array,
      default: () => []
    },
    disabled: {
      type: Boolean,
      default: false
    },
    handlePreview: {
      type: Function,
      default: () => shared.NOOP
    },
    listType: {
      type: String,
      default: "text"
    }
  },
  emits: ["remove"],
  setup(props, { emit }) {
    const { t } = index$2.useLocaleInject();
    const handleClick = (file) => {
      props.handlePreview(file);
    };
    const onFileClicked = (e) => {
      ;
      e.target.focus();
    };
    const handleRemove = (e, file) => {
      emit("remove", file);
    };
    return {
      focusing: vue.ref(false),
      handleClick,
      handleRemove,
      onFileClicked,
      t
    };
  }
});

exports["default"] = script;
