// _element-plus@1.2.0-beta.4@element-plus/lib/components/tree-v2/src/tree-node.vue_vue_type_script_lang.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
var icons = require('@element-plus/icons');
var index = require('../../icon/index.js');
var index$1 = require('../../checkbox/index.js');
var treeNodeContent = require('./tree-node-content.js');
var virtualTree = require('./virtual-tree.js');

const DEFAULT_ICON = "caret-right";
var script = vue.defineComponent({
  name: "ElTreeNode",
  components: {
    ElIcon: index.ElIcon,
    CaretRight: icons.CaretRight,
    ElCheckbox: index$1.ElCheckbox,
    ElNodeContent: treeNodeContent["default"]
  },
  props: virtualTree.treeNodeProps,
  emits: virtualTree.treeNodeEmits,
  setup(props, { emit }) {
    const tree = vue.inject(virtualTree.ROOT_TREE_INJECTION_KEY);
    const indent = vue.computed(() => {
      var _a;
      return (_a = tree == null ? void 0 : tree.props.indent) != null ? _a : 16;
    });
    const icon = vue.computed(() => {
      var _a;
      return (_a = tree == null ? void 0 : tree.props.icon) != null ? _a : DEFAULT_ICON;
    });
    const handleClick = () => {
      emit("click", props.node);
    };
    const handleExpandIconClick = () => {
      emit("toggle", props.node);
    };
    const handleCheckChange = (value) => {
      emit("check", props.node, value);
    };
    const handleContextMenu = (event) => {
      var _a, _b, _c, _d;
      if ((_c = (_b = (_a = tree == null ? void 0 : tree.instance) == null ? void 0 : _a.vnode) == null ? void 0 : _b.props) == null ? void 0 : _c["onNodeContextmenu"]) {
        event.stopPropagation();
        event.preventDefault();
      }
      tree == null ? void 0 : tree.ctx.emit(virtualTree.NODE_CONTEXTMENU, event, (_d = props.node) == null ? void 0 : _d.data, props.node);
    };
    return {
      indent,
      icon,
      handleClick,
      handleExpandIconClick,
      handleCheckChange,
      handleContextMenu
    };
  }
});

exports["default"] = script;
