// element-plus/es/components/message/src/message.vue_vue_type_template_id_031967c2_lang.mjs
import { resolveComponent, openBlock, createBlock, Transition, withCtx, withDirectives, createElementVNode, normalizeClass, normalizeStyle, createCommentVNode, resolveDynamicComponent, renderSlot, createElementBlock, toDisplayString, Fragment, withModifiers, createVNode, vShow } from 'vue';

const _hoisted_1 = ["id"];
const _hoisted_2 = {
  key: 0,
  class: "el-message__content"
};
const _hoisted_3 = ["innerHTML"];
function render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_el_badge = resolveComponent("el-badge");
  const _component_el_icon = resolveComponent("el-icon");
  const _component_close = resolveComponent("close");
  return openBlock(), createBlock(Transition, {
    name: "el-message-fade",
    onBeforeLeave: _ctx.onClose,
    onAfterLeave: _cache[2] || (_cache[2] = ($event) => _ctx.$emit("destroy"))
  }, {
    default: withCtx(() => {
      var _a;
      return [
        withDirectives(createElementVNode("div", {
          id: _ctx.id,
          class: normalizeClass([
            "el-message",
            _ctx.type && !_ctx.icon ? `el-message--${_ctx.type}` : "",
            _ctx.center ? "is-center" : "",
            _ctx.showClose ? "is-closable" : "",
            _ctx.customClass
          ]),
          style: normalizeStyle(_ctx.customStyle),
          role: "alert",
          onMouseenter: _cache[0] || (_cache[0] = (...args) => _ctx.clearTimer && _ctx.clearTimer(...args)),
          onMouseleave: _cache[1] || (_cache[1] = (...args) => _ctx.startTimer && _ctx.startTimer(...args))
        }, [
          _ctx.repeatNum > 1 ? (openBlock(), createBlock(_component_el_badge, {
            key: 0,
            value: _ctx.repeatNum,
            type: (_a = _ctx.type) != null ? _a : "info",
            class: "el-message__badge"
          }, null, 8, ["value", "type"])) : createCommentVNode("v-if", true),
          _ctx.iconComponent ? (openBlock(), createBlock(_component_el_icon, {
            key: 1,
            class: normalizeClass(["el-message__icon", _ctx.typeClass])
          }, {
            default: withCtx(() => [
              (openBlock(), createBlock(resolveDynamicComponent(_ctx.iconComponent)))
            ]),
            _: 1
          }, 8, ["class"])) : createCommentVNode("v-if", true),
          renderSlot(_ctx.$slots, "default", {}, () => [
            !_ctx.dangerouslyUseHTMLString ? (openBlock(), createElementBlock("p", _hoisted_2, toDisplayString(_ctx.message), 1)) : (openBlock(), createElementBlock(Fragment, { key: 1 }, [
              createCommentVNode(" Caution here, message could've been compromised, never use user's input as message "),
              createElementVNode("p", {
                class: "el-message__content",
                innerHTML: _ctx.message
              }, null, 8, _hoisted_3)
            ], 2112))
          ]),
          _ctx.showClose ? (openBlock(), createBlock(_component_el_icon, {
            key: 2,
            class: "el-message__closeBtn",
            onClick: withModifiers(_ctx.close, ["stop"])
          }, {
            default: withCtx(() => [
              createVNode(_component_close)
            ]),
            _: 1
          }, 8, ["onClick"])) : createCommentVNode("v-if", true)
        ], 46, _hoisted_1), [
          [vShow, _ctx.visible]
        ])
      ];
    }),
    _: 3
  }, 8, ["onBeforeLeave"]);
}

export { render };
