// _element-plus@1.2.0-beta.4@element-plus/lib/components/drawer/src/index.vue_vue_type_script_lang.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');
var index = require('../../overlay/index.js');
require('../../dialog/index.js');
var index$1 = require('../../icon/index.js');
require('../../../directives/index.js');
var icons = require('@element-plus/icons');
var index$2 = require('../../../directives/trap-focus/index.js');
var dialog = require('../../dialog/src/dialog.js');
var useDialog = require('../../dialog/src/use-dialog.js');

var script = vue.defineComponent({
  name: "ElDrawer",
  components: {
    ElOverlay: index.ElOverlay,
    ElIcon: index$1.ElIcon,
    Close: icons.Close
  },
  directives: {
    TrapFocus: index$2["default"]
  },
  props: {
    ...dialog.dialogProps,
    direction: {
      type: String,
      default: "rtl",
      validator: (val) => {
        return ["ltr", "rtl", "ttb", "btt"].indexOf(val) !== -1;
      }
    },
    size: {
      type: [String, Number],
      default: "30%"
    },
    withHeader: {
      type: Boolean,
      default: true
    },
    modalFade: {
      type: Boolean,
      default: true
    }
  },
  emits: dialog.dialogEmits,
  setup(props, ctx) {
    const drawerRef = vue.ref(null);
    return {
      ...useDialog.useDialog(props, ctx, drawerRef),
      drawerRef,
      isHorizontal: vue.computed(() => props.direction === "rtl" || props.direction === "ltr"),
      drawerSize: vue.computed(() => typeof props.size === "number" ? `${props.size}px` : props.size)
    };
  }
});

exports["default"] = script;
