// _element-plus@1.2.0-beta.4@element-plus/lib/components/message/src/message.vue_vue_type_template_id_031967c2_lang.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var vue = require('vue');

const _hoisted_1 = ["id"];
const _hoisted_2 = {
  key: 0,
  class: "el-message__content"
};
const _hoisted_3 = ["innerHTML"];
function render(_ctx, _cache, $props, $setup, $data, $options) {
  const _component_el_badge = vue.resolveComponent("el-badge");
  const _component_el_icon = vue.resolveComponent("el-icon");
  const _component_close = vue.resolveComponent("close");
  return vue.openBlock(), vue.createBlock(vue.Transition, {
    name: "el-message-fade",
    onBeforeLeave: _ctx.onClose,
    onAfterLeave: _cache[2] || (_cache[2] = ($event) => _ctx.$emit("destroy"))
  }, {
    default: vue.withCtx(() => {
      var _a;
      return [
        vue.withDirectives(vue.createElementVNode("div", {
          id: _ctx.id,
          class: vue.normalizeClass([
            "el-message",
            _ctx.type && !_ctx.icon ? `el-message--${_ctx.type}` : "",
            _ctx.center ? "is-center" : "",
            _ctx.showClose ? "is-closable" : "",
            _ctx.customClass
          ]),
          style: vue.normalizeStyle(_ctx.customStyle),
          role: "alert",
          onMouseenter: _cache[0] || (_cache[0] = (...args) => _ctx.clearTimer && _ctx.clearTimer(...args)),
          onMouseleave: _cache[1] || (_cache[1] = (...args) => _ctx.startTimer && _ctx.startTimer(...args))
        }, [
          _ctx.repeatNum > 1 ? (vue.openBlock(), vue.createBlock(_component_el_badge, {
            key: 0,
            value: _ctx.repeatNum,
            type: (_a = _ctx.type) != null ? _a : "info",
            class: "el-message__badge"
          }, null, 8, ["value", "type"])) : vue.createCommentVNode("v-if", true),
          _ctx.iconComponent ? (vue.openBlock(), vue.createBlock(_component_el_icon, {
            key: 1,
            class: vue.normalizeClass(["el-message__icon", _ctx.typeClass])
          }, {
            default: vue.withCtx(() => [
              (vue.openBlock(), vue.createBlock(vue.resolveDynamicComponent(_ctx.iconComponent)))
            ]),
            _: 1
          }, 8, ["class"])) : vue.createCommentVNode("v-if", true),
          vue.renderSlot(_ctx.$slots, "default", {}, () => [
            !_ctx.dangerouslyUseHTMLString ? (vue.openBlock(), vue.createElementBlock("p", _hoisted_2, vue.toDisplayString(_ctx.message), 1)) : (vue.openBlock(), vue.createElementBlock(vue.Fragment, { key: 1 }, [
              vue.createCommentVNode(" Caution here, message could've been compromised, never use user's input as message "),
              vue.createElementVNode("p", {
                class: "el-message__content",
                innerHTML: _ctx.message
              }, null, 8, _hoisted_3)
            ], 2112))
          ]),
          _ctx.showClose ? (vue.openBlock(), vue.createBlock(_component_el_icon, {
            key: 2,
            class: "el-message__closeBtn",
            onClick: vue.withModifiers(_ctx.close, ["stop"])
          }, {
            default: vue.withCtx(() => [
              vue.createVNode(_component_close)
            ]),
            _: 1
          }, 8, ["onClick"])) : vue.createCommentVNode("v-if", true)
        ], 46, _hoisted_1), [
          [vue.vShow, _ctx.visible]
        ])
      ];
    }),
    _: 3
  }, 8, ["onBeforeLeave"]);
}

exports.render = render;
